hazelcast/src/main/java/com/hazelcast/client/impl/protocol/ClientExceptionFactory.java

```diff
@@ -612,20 +612,40 @@ public Throwable createException(ClientMessage clientMessage) {
     public ClientMessage createExceptionMessage(Throwable throwable) {
         int errorCode = getErrorCode(throwable);
         String message = throwable.getMessage();
-        StackTraceElement[] stackTrace = throwable.getStackTrace();
-        Throwable cause = throwable.getCause();
-        boolean hasCause = cause != null;
-        String className = throwable.getClass().getName();
-        if (hasCause) {
-            int causeErrorCode = getErrorCode(cause);
-            String causeClassName = cause.getClass().getName();
-            return ErrorCodec.encode(errorCode, className, message, stackTrace,
-                    causeErrorCode, causeClassName);
-        } else {
-            return ErrorCodec.encode(errorCode, className, message, stackTrace,
-                    ClientProtocolErrorCodes.UNDEFINED, null);
+
+        // combine the stack traces of causes recursively into one long stack trace
+        // first, count total number of stack trace elements
+        StackTraceElement[] combinedStackTrace;
+        int stackTraceElementsTotal = 0;
+        Throwable t = throwable;
+        while (t != null) {
+            // +1 for separator
+            stackTraceElementsTotal += t.getStackTrace().length + 1;
+            t = t.getCause();
         }
+        // -1 for last separator
+        stackTraceElementsTotal--;
 
+        // then copy arrays and separators
+        combinedStackTrace = new StackTraceElement[stackTraceElementsTotal];
+        t = throwable;
+        int pos = 0;
+        while (t != null) {
+            StackTraceElement[] stackTrace = t.getStackTrace();
+            System.arraycopy(stackTrace, 0, combinedStackTrace, pos, stackTrace.length);
+            pos += stackTrace.length;
+            t = t.getCause();
+            // add separator, if there is one more cause
+            if (t != null) {
+                combinedStackTrace[pos] = new StackTraceElement("###### Caused by: " + t + " ------", "", null, -1);
+                pos++;
+            }
+        }
+
+        // encode the exception as if without cause
```

Comment:
I did not get why we are not keeping the error code of cause, and its class name.  There are exception factories that are using cause exception. Honestly, I am surprised that tests are passed without this. I was sure that at least one test is checking the cause class type, besides the `ClientExceptionFactoryTest`.